plugins/outputs/azuremonitor/azuremonitor.go

```diff
@@ -0,0 +1,445 @@
+package azuremonitor
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+var _ telegraf.AggregatingOutput = (*AzureMonitor)(nil)
+var _ telegraf.Output = (*AzureMonitor)(nil)
+
+// AzureMonitor allows publishing of metrics to the Azure Monitor custom metrics service
+type AzureMonitor struct {
+	useMsi              bool              `toml:"use_managed_service_identity"`
+	ResourceID          string            `toml:"resource_id"`
+	Region              string            `toml:"region"`
+	Timeout             internal.Duration `toml:"Timeout"`
+	AzureSubscriptionID string            `toml:"azure_subscription"`
+	AzureTenantID       string            `toml:"azure_tenant"`
+	AzureClientID       string            `toml:"azure_client_id"`
+	AzureClientSecret   string            `toml:"azure_client_secret"`
+	StringAsDimension   bool              `toml:"string_as_dimension"`
+
+	url         string
+	msiToken    *msiToken
+	oauthConfig *adal.OAuthConfig
+	adalToken   adal.OAuthTokenProvider
+
+	client *http.Client
+
+	cache map[time.Time]map[uint64]*aggregate
+}
+
+type aggregate struct {
+	telegraf.Metric
+	updated bool
+}
+
+const (
+	defaultRegion          string = "eastus"
+	defaultMSIResource     string = "https://monitoring.azure.com/"
+	urlTemplate            string = "https://%s.monitoring.azure.com%s/metrics"
+	resourceIDTemplate     string = "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/virtualMachines/%s"
+	vmInstanceMetadataURL  string = "http://169.254.169.254/metadata/instance?api-version=2017-12-01"
+	msiInstanceMetadataURL string = "http://169.254.169.254/metadata/identity/oauth2/token"
+)
+
+var sampleConfig = `
+  ## The resource ID against which metric will be logged.  If not
+  ## specified, the plugin will attempt to retrieve the resource ID
+  ## of the VM via the instance metadata service (optional if running 
+  ## on an Azure VM with MSI)
+  #resource_id = "/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.Compute/virtualMachines/<vm_name>"
+  ## Azure region to publish metrics against.  Defaults to eastus.
+  ## Leave blank to automatically query the region via MSI.
+  #region = "useast"
+
+  ## Write HTTP timeout, formatted as a string.  If not provided, will default
+  ## to 5s. 0s means no timeout (not recommended).
+  # timeout = "5s"
+
+  ## Whether or not to use managed service identity.
+  #use_managed_service_identity = true
```

Comment:
Nit: In the sample configuration, we should make it clear that when use_managed_service_identify is false, it's required that the user supply resource_id, region, azure_subscription, azure_tenant, azure_client_id, and azure_client_secret.  These become mandatory parameters.